### ISSUE TITLE
Jenkins Controllers in Azure VMs: use workload identity management to allow managing Azure VM agents without credentials

### ISSUE THREAD

**Issue by dduportal** [jenkins-infra/helpdesk#3460]
### Service(s)

cert.ci.jenkins.io, ci.jenkins.io, trusted.ci.jenkins.io

### Summary

Credential are annoying because they need to be rotated when expired and leads to service interruption.
Also, having credentials means having to store them encrypted and manage the safety of these persistent but sensitive data.

We had numerous issues wasting our time on this: https://github.com/jenkins-infra/helpdesk/issues/3395, https://github.com/jenkins-infra/helpdesk/issues/3459 etc..

The goal of this issue is to use [Azure Workload Identity](https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview) for these controllers to authenticate (instead of using a SP/App credential).

Controller with which to proceed:

- [x] cert.ci.jenkins.io
- [x] ci.jenkins.io
- [x] trusted.jenkins.io

Note: Workload identity can be used with AKS pods. Will be a subsequent issue: we focus on the combo "Azure VM Controller -> Azure VM agents" only, which excludes the following:
- infra.ci.jenkins.io (which uses Azure VM but runs inside a pod)
- release.ci.jenkins.io (which does not even use any Azure VM)

**Comment by timja** [comment]
> release.ci.jenkins.io (because it does not spawn any VM, but only containers)

? what's that got to do with workload identity. It can use it just fine

**Comment by dduportal** [comment]
> > release.ci.jenkins.io (because it does not spawn any VM, but only containers)
> 
> ? what's that got to do with workload identity. It can use it just fine

True, but the issue was written quickly. I've edited the body so it focuses only on "Azure VM Controller" + "Azure VM Agents".

The Workload identity for pods will be treated as a distinct scope.

**Comment by dduportal** [comment]
> Certbot plugin `dns-azure` does support workload identity: https://docs.certbot-dns-azure.co.uk/en/latest/#certbot-azure-workload-identity-ini

Looks like there is an issue with the support of workload identity in `certbot-dns-azure`, I've opened an issue: https://github.com/terricain/certbot-dns-azure/issues/55.

Currently trying with https://github.com/terricain/certbot-dns-azure which works (as I tested with `lego`) with:

```
# /etc/letsencrypt/dns-multi.ini
dns_multi_provider = azuredns
AZURE_TENANT_ID = "4c45fef2-1ba7-4120-80a0-9e2d03e9c2b6"
AZURE_RESOURCE_GROUP = "proddns_jenkinsio"
AZURE_ENVIRONMENT = "public"
AZURE_ZONE_NAME = "cert.ci.jenkins.io"
```

**Comment by dduportal** [comment]
> > Certbot plugin `dns-azure` does support workload identity: https://docs.certbot-dns-azure.co.uk/en/latest/#certbot-azure-workload-identity-ini
> 
> Looks like there is an issue with the support of workload identity in `certbot-dns-azure`, I've opened an issue: [terricain/certbot-dns-azure#55](https://github.com/terricain/certbot-dns-azure/issues/55).
> 
> Currently trying with https://github.com/alexzorin/certbot-dns-multi which works (as I tested with `lego`) with:
> 
> # /etc/letsencrypt/dns-multi.ini
> dns_multi_provider = azuredns
> AZURE_TENANT_ID = "4c45fef2-1ba7-4120-80a0-9e2d03e9c2b6"
> AZURE_RESOURCE_GROUP = "proddns_jenkinsio"
> AZURE_ENVIRONMENT = "public"
> AZURE_ZONE_NAME = "cert.ci.jenkins.io"

Implemented in https://github.com/jenkins-infra/jenkins-infra/pull/4086 (and a hotfix https://github.com/jenkins-infra/jenkins-infra/commit/bbd54d511cc9b9bf5ea344f9df2ae453235907bd)

**Comment by timja** [comment]
Did you try providing the client id?

**Comment by dduportal** [comment]
> Did you try providing the client id?

Yes, and it fails due to a missing federated token id. The `azure` function called by https://github.com/terricain/certbot-dns-azure/blob/5ebc2998da738b5a6f3847e9bedd3db768feb2cb/certbot_dns_azure/_internal/dns_azure.py#L151 seems to be the culprit:

- https://github.com/terricain/certbot-dns-azure/blob/5ebc2998da738b5a6f3847e9bedd3db768feb2cb/certbot_dns_azure/_internal/dns_azure.py#L12
- https://pypi.org/project/azure-identity/ seems to have distinct implementations for workload and managed identitites
